### PR TITLE
Validate Topology Cache Group non-emptiness on Delivery Services's CDN

### LINKED ISSUE #5303: Need to validate DS topology assignment (cachegroups must contain at least 1 server in the DS CDN)
## I'm submitting a ...

-  new feature / enhancement request

## Traffic Control components affected ...

-  Traffic Ops

## New behavior:

As a CDN operator, I would like TO to validate my chosen topology for my delivery service. My chosen topology should have at least one server in my delivery service's CDN per cachegroup in the topology, so that the topology provides a realistic view of my delivery service.

## Current behavior:

This validation does not yet exist.

## Minimal reproduction of the problem with instructions:

1. Make a Topology
2. Make a Delivery Service on CDN 1
3. Move all of the Topology's Cache Group's Servers to a different CDN
4. Assign the Topology to the Delivery Service anyway and watch the PUT succeed

### PR DESCRIPTION
**Pull request by zrhoffman**
## What does this PR (Pull Request) do?

- [x] This PR fixes #5303

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

* Run Traffic Ops API tests
* Make sure the Traffic Ops API tests have good coverage

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] Added validation and not new features, no documentation necessary
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 1]
```diff
@@ -839,18 +842,50 @@ func TopologyExists(tx *sql.Tx, name string) (bool, error) {
...
+	}
+
+	if err = topology_validation.CheckForEmptyCacheGroups(tx, cacheGroupIDs, []int{*ds.CDNID}, true, []int{}); err != nil {
+		return http.StatusBadRequest, fmt.Errorf("empty cachegroups in Topology %s found for CDN %d: %s", *ds.Topology, ds.CDNID, err.Error()), nil
+	}
+
+	return statusCode, userErr, sysErr
+}
+
+// GetTopologyCachegroups returns a map of cachegroup names by cachegroup id for the given topology.
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:869]: nit: this comment is wrong

File: traffic_ops/testing/api/v3/deliveryservices_test.go [hunk 2]
```diff
@@ -603,6 +603,110 @@ func UpdateDeliveryServiceWithInvalidTopology(t *testing.T) {
...
+	if len(cdns) != expectedSize {
+		t.Fatalf("expected %d CDN with name %s but instead received %d CDNs", expectedSize, cdn1Name, len(cdns))
+	}
+	cdn1 := cdns[0]
+	const cacheGroupName = "dtrc1"
+	cachegroups, _, err := TOSession.GetCacheGroupsByQueryParamsWithHdr(url.Values{"name": {cacheGroupName}}, nil)
+	if err != nil {
+		t.Fatalf("getting Cache Group %s: %s", cacheGroupName, err.Error())
+	}
+	if len(cdns) != expectedSize {
```
Comment (rawlinp) [inline traffic_ops/testing/api/v3/deliveryservices_test.go:637]: I think you meant to check `len(cachegroups)` here

File: traffic_ops/testing/api/v3/deliveryservices_test.go [hunk 3]
```diff
@@ -603,6 +603,110 @@ func UpdateDeliveryServiceWithInvalidTopology(t *testing.T) {
...
+	_, _, err = TOSession.CopyProfile(profileCopy)
+	if err != nil {
+		t.Fatalf("copying Profile %s: %s", *server.Profile, err.Error())
+	}
+
+	profiles, _, err := TOSession.GetProfileByNameWithHdr(profileCopy.Name, nil)
+	if err != nil {
+		t.Fatalf("getting Profile %s: %s", profileCopy.Name, err.Error())
+	}
+	if len(servers.Response) != expectedSize {
```
Comment (rawlinp) [inline traffic_ops/testing/api/v3/deliveryservices_test.go:668]: I think you meant to check `len(profiles)` here

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 4]
```diff
@@ -23,6 +23,9 @@ import (
 	"database/sql"
 	"errors"
 	"fmt"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/topology/topology_validation"
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:26]: nit: these imports are grouped haphazardly

**Review by rawlinp — CHANGES_REQUESTED** [review]
code looks good, just some minor things

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go (hunk 4 above)
Comment (zrhoffman) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go]: Organized imports in 564d00922f

File: traffic_ops/testing/api/v3/deliveryservices_test.go (hunk 3 above)
Comment (zrhoffman) [reply traffic_ops/testing/api/v3/deliveryservices_test.go]: Checking `len(profiles)` in 94272ce568

File: traffic_ops/testing/api/v3/deliveryservices_test.go (hunk 2 above)
Comment (zrhoffman) [reply traffic_ops/testing/api/v3/deliveryservices_test.go]: Checking `len(cachegroups)` in 94272ce568

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go (hunk 1 above)
Comment (zrhoffman) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go]: Made comment right in 4fab50ff5e

**Review by rawlinp — APPROVED** [review]

**Pull request merged**